ui/component/core/src/index.ts

```diff
@@ -1105,76 +962,118 @@ export class Manager implements EventProviderFactory {
-        if(this._eventsDisconnected !== disconnected || force) {
-            console.debug(`EventProvider status changed: ${disconnected ? 'DISCONNECTED' : 'CONNECTED'}`);
-            if(disconnected) {
-                this._eventsDisconnected = true;
-                this._emitEvent(OREvent.OFFLINE);
-
-                // WebSocket might not connect because of token expiry, so we try updating it immediately.
-                // Since token expiry is unrelated to ONLINE/OFFLINE state, a successful token update would change _authDisconnected from FALSE to FALSE.
-                // So, if the token update is successful, we force update _authDisconnected.
-                if(this.isTokenExpired() && !this._authDisconnected) {
-                    this._tryUpdateAccessToken().then((offline) => {
-                        if(!offline) {
-                            this._setAuthDisconnected(false, true);
-                        }
-                    });
-                }
+    /**
+     * Checks keycloak is available and token is valid otherwise will redirect to login; also checks if event bus is
+     * online.
+     */
+    public async reconnect(reattemptDelayMillis: number = 3000) {
 
-            } else {
-                // When EventProvider is back online, check if authentication is disconnected and try to reauthenticate if necessary.
-                // Only emit the ONLINE event when both are connected.
-                this._eventsDisconnected = false;
-                if(this._authDisconnected) {
-                    this._runAuthReconnectTimer()
-                } else {
-                    this._emitEvent(OREvent.ONLINE);
+        if (!this._disconnected) {
+            return;
+        }
+
+        if (this._reconnectTimer) {
+            window.clearTimeout(this._reconnectTimer);
+            this._reconnectTimer = undefined;
+        }
+
+        const tryReconnect = async () => {
+            let keycloakOffline = !await this.isKeycloakReachable();
+
+            if (keycloakOffline) {
+                return false;
+            }
+
+            // Check if access token can be refreshed
+            try {
+                await this._updateKeycloakAccessToken();
+            } catch (e) {
+                // Try and use offline token if it is available
+                const offlineToken = await this._getNativeOfflineRefreshToken();
+                this._keycloak!.refreshToken = offlineToken;
+
+                try {
+                    await this._updateKeycloakAccessToken();
+                } catch (e) {
+                    this.login();
+                    return;
                 }
+                return true;
             }
+
+            // Check events
+            const eventsOffline = this.events && this.events.status === EventProviderStatus.CONNECTING;
+            // Force reconnect attempt now if needed
+            return !eventsOffline || await this.events?.connect();
```

Comment:
So if I understand correctly;
If the websocket has disconnected, and this `reconnect()` function is triggered,
it will always request keycloak a new token, no matter if the token is still valid?

The other way around (token update fail, but websocket still alive) is catched with `eventsOffline`,
so it would think it's worth applying for the token update as well.

Reply:
Not quite `updateKeycloakAccessToken()` will only get a new token if the existing token is going to expire in next 20s. 
Keycloak must be available and tokens valid for